### ISSUE TITLE
Calling read on a pipe fd gets stuck

### ISSUE THREAD

**Issue by spellr** [eventlet/eventlet#529]
- Python version 2.7.5
- eventlet version 0.24.1
- Reproducible: always
- Expected result: getting data from a pipe
- Real result: the call hangs

My application is calling `os.fdopen`, on a pipe, and gets a reader. The initialization code is [here](https://github.com/tornadoweb/tornado/blob/v4.5.3/tornado/platform/posix.py#L44)
Then it [calls](https://github.com/tornadoweb/tornado/blob/v4.5.3/tornado/platform/posix.py#L62) reader.read(). Python's socket.read function is called, and it calls eventlet's GreenPipe's `recv` function indefinitely, until an empty string is returned.
`recv` [calls](https://github.com/eventlet/eventlet/blob/master/eventlet/greenio/py2.py#L161) eventlet's `read`.
When there's data in the pipe, `read` gets and and returns, then gets called again.
Problem is when there's nothing in the pipe, `read` gets "Resource temporarily unavailable" from `__original_read__`, and [tries again](https://github.com/eventlet/eventlet/blob/master/eventlet/green/os.py#L43), until there's something in the pipe.

This causes reader.read to get stuck, and it never returns.

**Comment by nat-goodspeed** [comment]
What happens when the other end of the pipe is closed? Does that unstick it?

What happens if, without `eventlet`, you open a socket and call its `makefile()` method, and call `read()` on that? Doesn't that exhibit the same behavior?

The [documented behavior](https://docs.python.org/2/library/stdtypes.html#file.read) of `read()` with omitted size argument is to "read all data until EOF is reached." The behavior you describe sounds to me like a reasonable interpretation.

**Comment by spellr** [comment]
> What happens when the other end of the pipe is closed? Does that unstick it?

It's hard for me to check right now, but from looking at the code, it seems it will solve it.

> What happens if, without eventlet, you open a socket and call its makefile() method, and call read() on that? Doesn't that exhibit the same behavior?

The code without eventlet monkey patching works fine. The read returns immediately with everything in the pipe right now.

> The documented behavior of read() with omitted size argument is to "read all data until EOF is reached." 
> The behavior you describe sounds to me like a reasonable interpretation.

I don't agree this is the right behavior. If you'll see cpython's [implementation ](https://github.com/python/cpython/blob/2.7/Lib/socket.py#L360) of socket's read, you'll see that it stops once there nothing in the socket.
Also, from `pipe(7)` [man-page](http://man7.org/linux/man-pages/man7/pipe.7.html):

> A pipe has a limited capacity.  If the pipe is full, then a write(2) will block or fail, depending on whether the O_NONBLOCK flag is set (see below).  Different implementations have different limits for the pipe capacity.  Applications should not rely on a particular capacity: an application should be designed so that a reading process consumes data as soon as it is available, so that a writing process does not remain blocked.

**Comment by nat-goodspeed** [comment]
I still think the present behavior is "a reasonable interpretation" of mapping `read()` with omitted `size` onto a socket. I won't dispute that `socket.makefile().read()` matches your expectations better, but one could argue that an application expecting `read()` to obtain _everything_ from a given file handle until `close()` would get a nasty (timing-dependent) surprise.

Pipes and TCP sockets have no natural delimiters other than `close()`. It's typical to read a fixed-size buffer and impose your own protocol on that basis.

Bottom line, though -- unless you can demonstrate that this is a recent change from earlier versions of `eventlet`, we risk breaking existing code that _relies_ on the implemented behavior.

**Comment by spellr** [comment]
I get what you're saying about breaking existing behavior, it's just that I don't think you'll find any place where the behavior is what you're suggesting.
That's why I think it's a bug, and why I think it contradicts the expectation of any networking programmer that'll want to use this function, knowing the behavior from other places.
I also know this behavior is already causing [bugs](https://bugs.launchpad.net/neutron/+bug/1802736) to happen in other places.

**Comment by spellr** [comment]
Also, the fd is configured to be non-blocking. With the behavior you're suggesting, the read() call is blocking. So this doesn't match expectation when [setting](https://github.com/tornadoweb/tornado/blob/v4.5.3/tornado/platform/posix.py#L34) the socket as `O_NONBLOCK`.

**Comment by spellr** [comment]
> Just so it's perfectly clear what are we talking about, can you provide a minimal code example to reproduce the issue?

Sure thing
```python
import os
import fcntl
import eventlet

def _set_nonblocking(fd):
    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
    fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)

eventlet.monkey_patch() # without this line - the script works fine

r, w = os.pipe()
_set_nonblocking(r)
reader = os.fdopen(r, "rb", 0)
writer = os.fdopen(w, "wb", 0)

writer.write(b"x")
reader.read() # hangs!
```

**Comment by spellr** [comment]
ping?

**Comment by nat-goodspeed** [comment]
I won't attempt to speak for @jstasiak about the code example you posted; he may still have some comments about your response to his earlier request.

But I do agree with his remark that

> Eventlet exposes a blocking interface on top of [non-blocking socket I/O]

You say

> I get what you're saying about breaking existing behavior, it's just that I don't think you'll find any place where the behavior is what you're suggesting.

What about a simple attempt to fetch the entire content of a web page? What if the server side writes that content in chunks, and there's a brief transmission delay between two chunks?

As I noted earlier, the one delimiter in socket communication that _is_ deterministic is EOF. I _do_ think that changing the behavior as you request would break existing code.
